### PR TITLE
Add a ProxyProtocol sample

### PR DESCRIPTION
**Pull request by Tratcher**
This is a sample for parsing the Proxy Protocol as a Kestrel connection middleware. The protocol allows level 4 proxies to forward information as a connection preface. The preface needs to be removed before HTTP parsing can happen.

We're publishing this to labs because multiple parties have expressed interest in it. We know it's not exhaustive, it (e.g. it only handles IPv4 addresses), but it's a good starting point for interested parties.

See https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt Section 2.2

Azure private link is one such proxy that sends this data:
https://docs.microsoft.com/en-us/azure/private-link/private-link-service-overview#properties (EnableProxyProtocol)

Ngnix has a first class implementation of this spec (inbound). https://docs.nginx.com/nginx/admin-guide/load-balancer/using-proxy-protocol/#configuring-nginx-to-accept-the-proxy-protocol

### CODE REVIEW

File: src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+                            // It is PPv2, and we have enough data for PPv2
+                            break;
+                        }
+
+                        // It is PPv2, and we don't have enough data for PPv2
+                    }
+                }
+                finally
+                {
+                    input.AdvanceTo(buffer.Start);
```
Comment (halter73) [inline src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs:79]: If this receives a partial proxy preamble, it could spin indefinitely.

```suggestion
                    input.AdvanceTo(buffer.Start, buffer.End);
```

https://docs.microsoft.com/en-us/dotnet/standard/io/pipelines#pipereader-common-problems

File: src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            while (true)
+            {
+                var result = await input.ReadAsync();
+                var buffer = result.Buffer;
+
+                try
+                {
+                    if (result.IsCompleted)
+                    {
+                        return;
```
Comment (halter73) [inline src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs:49]: Does Proxy Protocol v2 support half-open connections? I assume not. But if it does, we should inspect the buffer before exiting due to a FIN.

File: src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs (hunk 1 above)
Comment (halter73) [reply src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs:79]: This also means we'll probably have to have a loop inside the try/finally that calls ExtractPpv2Data to ensure we consume everything we can before saying we've examined the entire buffere.

Comment (halter73) [reply src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs:79]: I now see why we didn't want to advance the examined cursor:

1. We don't want to examine past the "Ppv2Data" for proxy protocol connections.
2. We don't want to examine anything in case IsProxyProtocol later returns false in a later iteration. 

If we cannot either extract all the Ppv2Data or determine that it's not a PPv2 connection after the first read, I think we're going to have to create a new IDuplexPipe that replaces the original Reader with a new one that doesn't have everything pre-examined and set it on ConnectionContext.Transport.

File: src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+                    input.AdvanceTo(buffer.Start);
+                }
+            }
+
+            await next();
+        }
+
+        private static bool IsProxyProtocol(byte[] buffer)
+        {
+            if (buffer.Length > 12)
```
Comment (davidfowl) [inline src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs:88]: Preamble.AsSpan().SequenceEqual(buffer)

File: src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.Extensions.Logging;
+
+namespace ProxyProtocol.Sample
+{
+    public static class ProxyProtocol
+    {
+        // The proxy protocol marker.
+        private static readonly byte[] Preamble = { 0x0D, 0x0A, 0x0D, 0x0A, 0x00, 0x0D, 0x0A, 0x51, 0x55, 0x49, 0x54, 0x0A };
```
Comment (davidfowl) [inline src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs:15]: This should be types as `ReadOnlySpan<byte>`

private static readonly ReadOnlySpan<byte> Preamble => new byte[] { 0x0D, 0x0A, 0x0D, 0x0A, 0x00, 0x0D, 0x0A, 0x51, 0x55, 0x49, 0x54, 0x0A };

Comment (Tratcher) [reply src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs:15]: `private static ReadOnlySpan<byte> Preamble => new byte[] { 0x0D, 0x0A, 0x0D, 0x0A, 0x00, 0x0D, 0x0A, 0x51, 0x55, 0x49, 0x54, 0x0A };`

File: src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs (hunk 2 above)
Comment (Tratcher) [reply src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs:49]: PPv2 is only a connection preface, the connection state isn't relevant to the protocol.

I don't think half-open is something we care about when the target audience is Kestrel HTTP scenarios.

File: src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs (hunk 1 above)
Comment (Tratcher) [reply src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs:79]: I think we can advance examined so long as it's a partial match, and if it's not a partial match we can break out.

**Comment by Tratcher** [comment]
@halter73 do you have time now to review this, or should we give it one more week?

File: src/ProxyProtocol/Client.Sample/Program.cs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            Stream stream = client.GetStream();
+
+            if (serverAddress.Scheme == Uri.UriSchemeHttps)
+            {
+                Console.WriteLine("Negotiating TLS");
+                var sslStream = new SslStream(stream);
+                await sslStream.AuthenticateAsClientAsync(new SslClientAuthenticationOptions()
+                {
+                    TargetHost = serverAddress.Host,
+                    RemoteCertificateValidationCallback = (_, __, ___, ____) => true,
```
Comment (halter73) [inline src/ProxyProtocol/Client.Sample/Program.cs:38]: Should we be disabling normal cert validation in a sample? Trusing dev certs should be good enough, right?

File: src/ProxyProtocol/Client.Sample/Program.cs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+                stream = sslStream;
+            }
+
+            Console.WriteLine("Sending message");
+            var data = CreateMessage(serverAddress);
+            for (var i = 0; i < data.Length; i++)
+            {
+                Console.Write(".");
+                await stream.WriteAsync(data, i, 1);
+                await Task.Delay(TimeSpan.FromSeconds(0.1));
```
Comment (halter73) [inline src/ProxyProtocol/Client.Sample/Program.cs:50]: What's the purpose of sending the data one byte at a time?

File: src/ProxyProtocol/Client.Sample/Program.cs [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        {
+            var serverAddress = new Uri("https://localhost:5001");
+
+            Console.WriteLine("Ready");
+            Console.ReadKey();
+
+            await ConnectAsync(serverAddress);
+        }
+
+        private static async Task ConnectAsync(Uri serverAddress)
```
Comment (halter73) [inline src/ProxyProtocol/Client.Sample/Program.cs:23]: This is doing more than connecting. It looks like this is the entire client sample. Maybe just move this into `Main` or rename it to something like `ConnectAndMakeSampleRequestAsync`.

File: src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+                        // Break if it is not PPv2.
+                        break;
+                    }
+
+                    if (HasEnoughPpv2Data(bufferArray))
+                    {
+                        // It is PPv2
+                        ExtractPpv2Data(ref buffer, bufferArray, connectionContext, logger);
+                        // We've consumed and sliced off the prefix.
+                        examined = buffer.Start;
```
Comment (halter73) [inline src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs:84]: This could move examined backwards relative to a previous call to AdvanceTo. This will cause the default PipeReader to throw as it tries to get the length of newly-examined bytes.

https://github.com/dotnet/runtime/blob/00d0dba14ed9927b22de224de94d812449529f17/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs#L461

Comment (halter73) [reply src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs:84]: Can we just move examined to the `16 + length` evaluated in HasEnoughPpv2Data? I assume the TLS handshake cannot complete until there's more data than that.

File: src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+                var srcAddress = new IPAddress(srcIpAddressArray);
+
+                var destIpAddressArray = new byte[4];
+                Array.Copy(bufferArray, 20, destIpAddressArray, 0, 4);
+                var destAddress = new IPAddress(destIpAddressArray);
+
+                var srcPort = (int)(bufferArray[25] | (bufferArray[24] << 8));
+                var destPort = (int)(bufferArray[27] | (bufferArray[26] << 8));
+
+                context.Items.Add(SourceIPAddressKey, srcAddress);
```
Comment (davidfowl) [inline src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs:117]: Don't use Items here, use a single feature with 4 fields.

File: src/ProxyProtocol/Client.Sample/Program.cs (hunk 6 above)
Comment (davidfowl) [reply src/ProxyProtocol/Client.Sample/Program.cs:50]: This looks like a test 😄

Comment (Tratcher) [reply src/ProxyProtocol/Client.Sample/Program.cs:50]: To test boundary conditions 😁. And yes, this is a manual test.

File: src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs (hunk 8 above)
Comment (Tratcher) [reply src/ProxyProtocol/ProxyProtocol.Sample/ProxyProtocol.cs:84]: It looks like a TLS hello is way more than our 12 byte prefix minimum so rounding up shouldn't be an issue. Note this could also be used without TLS, but I wouldn't expect an issue there either, even `GET / HTTP/1.1` is more than 12 bytes.

**Comment by Tratcher** [comment]
Updated and rebased.

**Review by halter73 — APPROVED** [review]
